lib/cn/other_iterator.c

```diff
@@ -0,0 +1,80 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2022 Micron Technology, Inc.  All rights reserved.
+ */
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <hse_util/element_source.h>
+#include <hse_util/event_counter.h>
+#include <hse_util/hse_err.h>
+
+#include "kvset.h"
+#include "kvset_internal.h"
+#include "omf.h"
+
+struct other_iterator {
+    struct kvset *oi_ks;
+    struct element_source oi_es;
+    bool oi_eof;
+    uint32_t oi_offset;
+};
+
+static bool
+next(struct element_source *source, void **data)
+{
+    struct other_iterator *iter = container_of(source, struct other_iterator, oi_es);
+
+    if (iter->oi_eof)
+        return false;
+
+    assert(iter->oi_offset > 0);
+
+    *data = &iter->oi_ks->ks_kblks[--iter->oi_offset];
+
+    if (iter->oi_offset == 0)
+        iter->oi_eof = true;
+
+    return iter->oi_eof;
+}
+
+merr_t
+other_iterator_create(
+    struct kvset *const ks,
+    struct other_iterator **const iter)
+{
+    struct other_iterator *tmp;
+
+    *iter = NULL;
+
+    tmp = calloc(1, sizeof(*tmp));
+    if (ev(!tmp))
+        return merr(ENOMEM);
+
+    tmp->oi_ks = ks;
+    /* Using prefix decrement in next, so this is fine */
+    tmp->oi_offset = ks->ks_st.kst_kblks;
+    tmp->oi_es = es_make(next, NULL, NULL);
+
+    *iter = tmp;
+
+    return 0;
+}
+
+void
+other_iterator_destroy(struct other_iterator *const iter)
+{
+    if (!iter)
+        return;
+
+    free(iter);
+}
+
+struct element_source *
+other_iterator_get_element_source(struct other_iterator *const iter)
+{
+    return &iter->oi_es;
+}
```

Comment:
Always confuses me..  is *iter const, or is iter const?  if the former shouldn't we return a ptr to a const element_source?

Reply:
the pointer is const. I can't reassign iter, but I can mutate it.